server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/index/StarTreeValues.java

```diff
@@ -0,0 +1,173 @@
+
+        StarTreeMetadata starTreeMetadata = (StarTreeMetadata) compositeIndexMetadata;
+
+        // build skip star node dimensions
+        Set<String> skipStarNodeCreationInDims = starTreeMetadata.getSkipStarNodeCreationInDims();
+
+        // build dimensions
+        List<Dimension> readDimensions = new ArrayList<>();
+        for (String dimension : starTreeMetadata.getDimensionFields()) {
+            readDimensions.add(new ReadDimension(dimension));
+        }
+
+        // star-tree field
+        this.starTreeField = new StarTreeField(
+            starTreeMetadata.getCompositeFieldName(),
+            readDimensions,
+            starTreeMetadata.getMetrics(),
+            new StarTreeFieldConfiguration(
+                starTreeMetadata.getMaxLeafDocs(),
+                skipStarNodeCreationInDims,
+                starTreeMetadata.getStarTreeBuildMode()
+            )
+        );
+
+        this.root = StarTreeFactory.createStarTree(compositeIndexIn, starTreeMetadata);
+
+        // get doc id set iterators for metrics and dimensions
+        dimensionDocValuesIteratorMap = new LinkedHashMap<>();
+        metricDocValuesIteratorMap = new LinkedHashMap<>();
+
+        // get doc id set iterators for dimensions
+        for (String dimension : starTreeMetadata.getDimensionFields()) {
+            SortedNumericDocValues dimensionSortedNumericDocValues = null;
+            if (readState != null) {
+                FieldInfo dimensionfieldInfo = readState.fieldInfos.fieldInfo(
+                    fullyQualifiedFieldNameForStarTreeDimensionsDocValues(starTreeField.getName(), dimension)
+                );
+                if (dimensionfieldInfo != null) {
+                    dimensionSortedNumericDocValues = compositeDocValuesProducer.getSortedNumeric(dimensionfieldInfo);
+                }
+            }
+
+            dimensionDocValuesIteratorMap.put(dimension, getSortedNumericDocValues(dimensionSortedNumericDocValues));
+        }
+
+        // get doc id set iterators for metrics
+
+        for (Metric metric : starTreeMetadata.getMetrics()) {
+            for (MetricStat metricStat : metric.getMetrics()) {
+                String metricFullName = fullyQualifiedFieldNameForStarTreeMetricsDocValues(
+                    starTreeField.getName(),
+                    metric.getField(),
+                    metricStat.getTypeName()
+                );
+
+                SortedNumericDocValues metricSortedNumericDocValues = null;
+                if (readState != null) {
+                    FieldInfo metricFieldInfo = readState.fieldInfos.fieldInfo(metricFullName);
+                    if (metricFieldInfo != null) {
+                        metricSortedNumericDocValues = compositeDocValuesProducer.getSortedNumeric(metricFieldInfo);
+                    }
+                }
+                metricDocValuesIteratorMap.put(metricFullName, getSortedNumericDocValues(metricSortedNumericDocValues));
+            }
+        }
```

Comment:
Can this be done lazily when we get call to actually get the values using a memoized supplier kind of interface?